BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -399,263 +399,328 @@ def browse
-    # search of what's registered to only those belonging to @as_user
-    registered_files = registered_files.where( :user_id => @as_user.id ) if ! @provider.allow_file_owner_change?
-    registered_files = registered_files.all.index_by(&:name)
+        rescue => e
+          (failed[e.message] ||= []) << basename
+        end
+      end
 
-    basenames.shuffle.each do |basename|
+      # Remind the user if browsing as another user
+      flash[:notice] += "Important note! Since you were browsing as user '#{@as_user.login}', the files were registered as belonging to that user instead of you!\n" if
+        succeeded.present? && @as_user != current_user
 
-      # Unregister files
+      # If files actually got registered, clear the browsing cache
+      clear_browse_provider_local_cache_file(@as_user, @provider) if
+        succeeded.present? && [:html, :js].include?(request.format.to_sym)
 
-      if do_unreg || do_erase
-        userfile = registered_files[basename]
-        if userfile.blank?
-          num_skipped += 1 unless do_erase
-        elsif ! userfile.has_owner_access?(current_user)
-          flash[:error] += "Error: file #{basename} is not registered such that you have the necessary permissions to unregister it. File not unregistered.\n"
-          num_skipped += 1
-          next
-        else
-          num_unregistered += Userfile.delete(userfile.id) # NOT destroy()! We don't want to delete the content!
-          userfile.destroy_log rescue true
-        end
-        next unless do_erase
+      # No need to move or copy? Just set the file sizes and exit.
+      unless post_action
+        registered.each { |userfile| userfile.set_size! rescue true }
+        generic_notice_messages('register', succeeded, failed)
+        next
       end
 
-      # Erase unregistered files
-
-      if do_erase
-        temp_class    = FileCollection   # erasing should work whether or not target really is a directory or not; if not change this
-        temp_userfile = temp_class.new(
-           :name          => basename,
-           :data_provider => @provider,
-           :user_id       => @as_user.id, # cannot use current_user, since it might be a vault_ssh dp
-           :group_id      => current_user.own_group.id
-        ).freeze # do not save this file! it's only used temporarily to delete the content on the DP
-        erase_ok = @provider.provider_erase(temp_userfile) rescue nil
-        if erase_ok
-          num_erased += 1
-        else
-          num_skipped += 1
-        end
-        next
+      # Prepare to copy/move the files to the new DP
+      generic_notice_messages('register', succeeded, failed)
+      succeeded, failed = [], {}
+
+      # Will some of the file names collide?
+      collisions = Userfile.where(
+        :name             => registered.map(&:name),
+        :user_id          => current_user.id,
+        :data_provider_id => target_dp.id
+      ).map(&:name)
```

Comment:
Change `map()` to `raw_first_column('userfiles.name')` for speed, especially if thousands of files involved.